### PR TITLE
Java frontend: treat CProver stubbed methods more like normal stubs

### PR DESCRIPTION
**Pull request by smowton**
We already tried to ignore org.cprover.CProver methods such as nondetInt, relying upon
the front-end to supply stub bodies if desired based on their type. However, by creating
symbols for them as for normal (non-stub) methods, but then refusing to provide a body or
advertise them as normal methods, we failed to set up their type properly (specifically
the parameter names), as they were neither set like a stub (upon symbol creation, upon
discovery by a caller) or like a normal method (during java_bytecode_convert_method). By
refusing to create the symbol at all, this treats the special CProver methods more like
normal stubs: their symbols are now created on demand.

It is also no longer necessary for java_bytecode_languaget::methods_provided to special
case them, as they are not added to method_bytecodet either, again matching "real" stub
methods more closely.

### CODE REVIEW

File: jbmc/regression/jbmc/fake_stubs/test.desc [hunk 1]
```diff
@@ -0,0 +1,11 @@
+CORE
+Test.class
+--function Test.main --cp `../../../../scripts/format_classpath.sh . ../../../lib/java-models-library/target/core-models.jar`
+^VERIFICATION SUCCESSFUL$
+^EXIT=0$
+^SIGNAL=0$
+--
+--
+This calls to CProver.startThread, which are special-cased by the Java front-end (which omits the method's body
```
Comment (tautschnig) [inline jbmc/regression/jbmc/fake_stubs/test.desc]: nit-pick: s/are/is/

File: jbmc/regression/jbmc/fake_stubs/test.desc [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+in hope of the front-end handling it like a stub), but which don't get their parameter names assigned by the usual
```
Comment (tautschnig) [inline jbmc/regression/jbmc/fake_stubs/test.desc]: s/don't/doesn't/  s/their/its/

**Review by tautschnig — APPROVED** [review]

File: jbmc/src/java_bytecode/java_bytecode_convert_method.cpp [hunk 3]
```diff
@@ -406,8 +415,9 @@ void java_bytecode_convert_methodt::convert(
   // (e.g. "my.package.ClassName.myMethodName:(II)I") and query the symbol table
   // to retrieve the symbol (constructed by java_bytecode_convert_method_lazy)
   // associated to the method
-  const irep_idt method_identifier=
-    id2string(class_symbol.name)+"."+id2string(m.name)+":"+m.descriptor;
+  const irep_idt method_identifier =
```
Comment (thk123) [inline jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:418]: I mean not your code, but why do we have a local variable `method_identifier` that then is immediately assigned into `method_id`, perhaps just directly assign?

**Review by thk123 — APPROVED** [review]
I mean fine, but this seems like a messy approach and I don't fully understand it. Why not just not load these methods (however it is they are inviting the front ends to ignore them?)

**Comment by smowton** [comment]
@tautschnig @thk123 that solution didn't work for symex-driven-lazy-loading, so I completely changed my approach; see updated description.

**Review by thk123 — APPROVED** [review]
Orders of magnitude better I'd say :tada:

File: jbmc/src/java_bytecode/java_bytecode_convert_class.cpp [hunk 4]
```diff
@@ -154,11 +155,17 @@ class java_bytecode_convert_classt:public messaget
...
   /// \return true if the method is an ignored method, else false
-  static bool is_ignored_method(const methodt &method)
+  static bool is_ignored_method(
+    const irep_idt &class_name, const methodt &method)
   {
-    return method.has_annotation(ID_ignored_method);
+    static irep_idt org_cprover_CProver_name = "org.cprover.CProver";
+    return
+      (class_name == org_cprover_CProver_name &&
+       cprover_methods_to_ignore.count(id2string(method.name))) ||
```
Comment (tautschnig) [inline jbmc/src/java_bytecode/java_bytecode_convert_class.cpp]: nit-pick: Visual Studio would like to have an explicit `!= 0` with `.count(...)`.

**Review by tautschnig — APPROVED** [review]

**Pull request merged**